### PR TITLE
3DS touchscreen controls toggle

### PR DESCRIPTION
**Pull request by troybonneau**
Hello! This PR intends to replace the current keyboard on the 3DS touch screen with one that can be toggled. This is accomplished by adding an additional row of buttons and shifting the current interface down a bit. I have this marked WIP because while it is compiling fine I think I am missing a step to get .3dsx files working with the homebrew launcher because it crashes to a black screen on startup. This is my first PR here so any guidance is appreciated! 😄 

keyboard_on

### CODE REVIEW

**Comment by carstene1ns** [comment]
Good work so far.

What I dislike here is the big top row for just disabling the screen. IMO this should be made smaller, so the others buttons only loose a bit of height, I would suggest around 20-26 pixels.
This complicates the button logic a bit, though.
I have an old branch for this, but unfinished... The top row has additional space for example for a battery indicator and a restart(<kbd>F11</kbd>) button.

Code for the battery indicator (drawing in whole, fixed image should be better):
```c++
// in anonymous namespace
u8 battery_charging = 0;
u8 battery_level = 0;
u32 c2d_black = C2D_Color32f(0, 0, 0, 1);
u32 c2d_gray = C2D_Color32f(0.8f, 0.8f, 0.8f, 1);
u32 c2d_white = C2D_Color32f(1, 1, 1, 1);
u32 c2d_red = C2D_Color32f(1, 0, 0, 1);
u32 c2d_orange = C2D_Color32f(0, 1, 1, 1);
u32 c2d_green = C2D_Color32f(0, 1, 0, 1);

// in constructor
ptmuInit();

// in destructor
ptmuExit();

// in processEvents()
PTMU_GetBatteryChargeState(&battery_charging);
PTMU_GetBatteryLevel(&battery_level);

// in UpdateDisplay() when drawing on bottom screen
// battery indicator
int battery_x = 296;
int battery_y = 6;
u32 color = c2d_gray;
if (battery_charging) {
    color = c2d_white;
}
C2D_DrawRectSolid(battery_x, battery_y, 0.5f, 18, 8, color);
C2D_DrawRectSolid(battery_x + 18, battery_y + 3, 0.5f, 2, 2, color);
if (battery_level > 4) color = c2d_green;
else if (battery_level > 2) color = c2d_orange;
else color = c2d_red;
for (int i = 0; i < battery_level; i++)
    C2D_DrawRectSolid(battery_x + 2 + i*2 + i, battery_y + 2, 0.5f, 2, 4, color);
```

**Comment by Ghabry** [comment]
Reset is already mapped to KEY_SELECT.

I like the proposed solution of Schoko for now (the basic version in the middle)

When the battery indicator is ever added this can be still redesigned again (make everything smaller) in a new PR.

**Comment by troybonneau** [comment]
easyrpgkeyboard

I've adjusted the layout and code to work with the proposed basic version in the middle

**Comment by Ghabry** [comment]
I do not really understand why you created a new image instead of just taking the one from Schoko but okay...

Imo this is now good enough for inclusion, can you mark the PR as ready?

File: src/platform/3ds/3ds_ui.cpp [hunk 1]
```diff
@@ -211,15 +213,33 @@ void CtrUi::ProcessEvents() {
...
+
+			} else {
+				keys[keys_tbl[col + (row * 4)]] = true;
+			}
+
+			touch_x = pos.px;
+			touch_y = pos.py;
+		} else {
+			gspLcdInit();
+			if (R_SUCCEEDED(gspLcdInit())) {
```
Comment (carstene1ns) [inline src/platform/3ds/3ds_ui.cpp:237]: Remove first `gspLcdInit()`.

File: src/platform/3ds/3ds_ui.cpp [hunk 2]
```diff
@@ -211,15 +213,33 @@ void CtrUi::ProcessEvents() {
...
+		if(show_touchscreen) {
+			touch_state = 1;
+			touchPosition pos;
+			hidTouchRead(&pos);
+			u8 col = pos.px / button_width;
+			u8 row = pos.py / button_height;
+
+			if (row == 0 && col == 3) { // Turn off touchscreen for top right button
+				show_touchscreen = false;
+				gspLcdInit();
```
Comment (carstene1ns) [inline src/platform/3ds/3ds_ui.cpp:225]: Please also with error checking.

**Review by carstene1ns — CHANGES_REQUESTED** [review]

**Review by carstene1ns — APPROVED** [review]
Okay for now.

**Review by fdelapena — APPROVED** [review]

**Comment by Ghabry** [comment]
jenkins: test this please

File: src/platform/3ds/3ds_ui.cpp [hunk 3]
```diff
@@ -325,6 +325,7 @@ void CtrUi::UpdateDisplay() {
 		u8 row = touch_y / button_height;
 		u8 pos_x = col * button_width;
 		u8 pos_y = row * button_height;
+		u8 draw_width = button_width;
```
Comment (carstene1ns) [inline src/platform/3ds/3ds_ui.cpp:328]: instead simply replace it

**Comment by carstene1ns** [comment]
jenkins: ok to test

**Comment by carstene1ns** [comment]
Looks good to squash and merge 👍🏼

**Review by troybonneau — COMMENTED** [review]
Alright looks like everything is building well. Ready on my end.

**Pull request merged**